tempesta_fw/ss_skb.c

```diff
@@ -816,6 +816,50 @@ __skb_fragment(struct sk_buff *skb_head, struct sk_buff *skb, char *pspt,
 			ret = __split_pgfrag(skb_head, skb, i, offset, len, it, fragn);
 			goto done;
 		}
+		/*
+		 * In case another fragment does not exist, remember the
+		 * fragment number, after which you need to insert/delete data.
+		 */
+		if (offset == d_size)
+			j = i;
+	}
+
+	if (j >= 0) {
+		skb_frag_t *frag = &si->frags[j + 1];
+
+		if (len > 0) {
+			ret = __new_pgfrag(skb_head, skb, len, j + 1, 1);
+			__it_next_data(skb, j + 1, it, fragn);
+			goto done;
+		}
+
+		if (unlikely(-len > skb_frag_size(frag))) {
+			T_WARN("Attempt to delete too much\n");
+			return -EFAULT;
+		}
+
+		/* delete a full fragment. */
+		ret = 0;
+		if (unlikely(-len == skb_frag_size(frag))) {
+			ss_skb_adjust_data_len(skb, len);
+			__skb_frag_unref(frag);
+			if (j + 2 < si->nr_frags)
+				memmove(&si->frags[j + 1], &si->frags[j + 2],
+					(si->nr_frags - j - 2) * sizeof(skb_frag_t));
+			--si->nr_frags;
+			__it_next_data(skb, j + 1, it, fragn);
+			goto done;
+		}
+
+		/* delete the head part of a fragment. */
+		frag->page_offset -= len;
+		skb_frag_size_sub(frag, -len);
+		skb->len += len;
+		skb->data_len += len;
+		it->data = skb_frag_address(frag);
+		it->skb = skb;
+		*fragn = j + 1;
+		goto done;
```

Comment:
It seems this just equals to `__split_pgfrag_del_w_frag(skb_head, skb, j + 1, 0, -len, it, fragn)`